### PR TITLE
[Rule Tuning] Timeline Templates For Windows and Linux

### LINKED ISSUE #1879: [Rule Tuning] Timeline Templates For Windows and Linux
## Related
elastic/security-team/issues/2091

## Link to rule
Rules are TBD as this is a WIP. They will be added to this original comment.

## Description
Discovering and testing rules to have a timeline template.

[WiP Process Documentation](https://docs.google.com/document/d/1kGRlZf_2m4LxJT14PnKS1ejob8-zGwZV6kxJZvhYDiQ/edit#)

## References
[WIP Process](https://docs.google.com/document/d/1kGRlZf_2m4LxJT14PnKS1ejob8-zGwZV6kxJZvhYDiQ/edit?usp=sharing)

### PR DESCRIPTION
**Pull request by terrancedejesus**
## Issues
Resolves #1879 

## Summary
The following rules have been chosen to test these timeline templates with.

- [x] Hosts File Modified
- [x] Interactive Terminal Spawned via Python
- [x] Suspicious CertUtil Commands
- [x] RDP (Remote Desktop Protocol) from the Internet
- [x] LSASS Memory Dump Creation
- [x] Adding Hidden File Attribute via Attrib
- [x] Telnet Port Activity
- [x] ImageLoad via Windows Update Auto Update Client
- [x] Startup or Run Key Registry Modification
- [x] Uncommon Registry Persistence Change

## Process Tracking
All information regarding steps taken can be found in this [gdoc](https://docs.google.com/document/d/1kGRlZf_2m4LxJT14PnKS1ejob8-zGwZV6kxJZvhYDiQ/edit?usp=sharing)

### CODE REVIEW

**Comment by terrancedejesus** [comment]
## Hosts File Modified
Replicated by moving `hosts` file located at `C:\Windows\System32\drivers\etc\hosts` to `C:\Windows\System32\drivers\hosts`. 

### Comprehensive Timeline Query
Screen Shot 2022-03-28 at 11 53 46 AM

### Timeline Added to Duplicate Rule
Screen Shot 2022-03-28 at 11 37 50 AM

**Comment by terrancedejesus** [comment]
### Rule Tuning
While testing spawn shells with python, I noticed `Interactive Terminal Spawned via Python` only accounts for `python` as a process name and not `python3` which caused no alerts to generate. Updated in this branch since I am still tuning rules essentially.

<details>
<summary>Data</summary>
{
  "_index": ".ds-logs-endpoint.events.process-default-2022.03.28-000001",
  "_id": "Te9B238BDxJNEmTs_fNq",
  "_version": 1,
  "_score": 1,
  "_source": {
    "agent": {
      "id": "7bc5ce5c-de4a-4b79-81fa-45198c4d75b7",
      "type": "endpoint",
      "version": "8.2.0-SNAPSHOT"
    },
    "process": {
      "Ext": {
        "ancestry": [
          "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTEwMjY0LTEzMjkzMTI0NDk5LjQwMDAwMDAw",
          "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTEwMjYzLTEzMjkzMTI0NDk4Ljg5MDAwMDAwMA==",
          "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTEwMTg5LTEzMjkzMTI0NDk2Ljk1MDAwMDAwMA==",
          "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTg5Mi0xMzI5MzEyMTcyMS4w",
          "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTEtMTMyOTMxMjE3MDAuMA=="
        ]
      },
      "args": [
        "python3",
        "-c",
        "import pty; pty.spawn(\"/bin/sh\")"
      ],
      "parent": {
        "args": [
          "-bash"
        ],
        "name": "bash",
        "pid": 10264,
        "args_count": 0,
        "entity_id": "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTEwMjY0LTEzMjkzMTI0NDk5LjQwMDAwMDAw",
        "command_line": "-bash",
        "executable": "/bin/bash"
      },
      "exit_code": 0,
      "name": "python3",
      "pid": 10285,
      "args_count": 0,
      "entity_id": "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTEwMjg1LTEzMjkzMTI0NTkxLjUwMDAwMDAwMA==",
      "command_line": "python3 -c import pty; pty.spawn(\"/bin/sh\")",
      "executable": "/usr/bin/python3",
      "hash": {
        "sha1": "9b6d3f9d2129510da043a88f23c1d7dafdf8d104",
        "sha256": "2e833afd6114ff314d879486c01ec19d3d94ef6f8c808d193fb329c98c674097",
        "md5": "9822d4931aea8fb83f794d2d54c3c992"
      }
    },
    "message": "Endpoint process event",
    "@timestamp": "2022-03-30T14:36:38.2671779Z",
    "ecs": {
      "version": "1.11.0"
    },
    "data_stream": {
      "namespace": "default",
      "type": "logs",
      "dataset": "endpoint.events.process"
    },
    "elastic": {
      "agent": {
        "id": "7bc5ce5c-de4a-4b79-81fa-45198c4d75b7"
      }
    },
    "host": {
      "hostname": "ubuntu-tdejesus",
      "os": {
        "Ext": {
          "variant": "Ubuntu"
        },
        "kernel": "5.13.0-1019-gcp #23~20.04.1-Ubuntu SMP Mon Mar 7 13:39:50 UTC 2022",
        "name": "Linux",
        "family": "ubuntu",
        "type": "linux",
        "version": "20.04.4",
        "platform": "ubuntu",
        "full": "Ubuntu 20.04.4"
      },
      "ip": [
        "127.0.0.1",
        "::1",
        "10.142.0.52",
        "fe80::4001:aff:fe8e:34"
      ],
      "name": "ubuntu-tdejesus",
      "id": "c78e287bbb0a5328e28b435c97de1cd0",
      "mac": [
        "42:01:0a:8e:00:34"
      ],
      "architecture": "x86_64"
    },
    "event": {
      "agent_id_status": "verified",
      "sequence": 62669,
      "ingested": "2022-03-30T14:36:49Z",
      "created": "2022-03-30T14:36:38.2671779Z",
      "kind": "event",
      "module": "endpoint",
      "action": "end",
      "id": "MYFVxCublyiVcrVQ++++01D+",
      "category": [
        "process"
      ],
      "type": [
        "end"
      ],
      "dataset": "endpoint.events.process"
    },
    "user": {
      "Ext": {
        "real": {
          "id": 1022
        }
      },
      "id": 1022
    },
    "group": {
      "Ext": {
        "real": {
          "id": 1023
        }
      },
      "id": 1023
    }
  },
  "fields": {
    "process.hash.md5": [
      "9822d4931aea8fb83f794d2d54c3c992"
    ],
    "host.os.full.text": [
      "Ubuntu 20.04.4"
    ],
    "process.command_line.caseless": [
      "python3 -c import pty; pty.spawn(\"/bin/sh\")"
    ],
    "event.category": [
      "process"
    ],
    "process.name.text": [
      "python3"
    ],
    "host.os.name.text": [
      "Linux"
    ],
    "host.os.full": [
      "Ubuntu 20.04.4"
    ],
    "process.parent.command_line": [
      "-bash"
    ],
    "process.parent.name": [
      "bash"
    ],
    "process.parent.pid": [
      10264
    ],
    "process.hash.sha256": [
      "2e833afd6114ff314d879486c01ec19d3d94ef6f8c808d193fb329c98c674097"
    ],
    "host.hostname": [
      "ubuntu-tdejesus"
    ],
    "process.pid": [
      10285
    ],
    "host.mac": [
      "42:01:0a:8e:00:34"
    ],
    "elastic.agent.id": [
      "7bc5ce5c-de4a-4b79-81fa-45198c4d75b7"
    ],
    "process.parent.entity_id": [
      "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTEwMjY0LTEzMjkzMTI0NDk5LjQwMDAwMDAw"
    ],
    "host.os.version": [
      "20.04.4"
    ],
    "host.os.name": [
      "Linux"
    ],
    "host.name": [
      "ubuntu-tdejesus"
    ],
    "event.agent_id_status": [
      "verified"
    ],
    "event.kind": [
      "event"
    ],
    "user.id": [
      "1022"
    ],
    "host.os.type": [
      "linux"
    ],
    "process.Ext.ancestry": [
      "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTEwMjY0LTEzMjkzMTI0NDk5LjQwMDAwMDAw",
      "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTEwMjYzLTEzMjkzMTI0NDk4Ljg5MDAwMDAwMA==",
      "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTEwMTg5LTEzMjkzMTI0NDk2Ljk1MDAwMDAwMA==",
      "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTg5Mi0xMzI5MzEyMTcyMS4w",
      "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTEtMTMyOTMxMjE3MDAuMA=="
    ],
    "user.Ext.real.id": [
      "1022"
    ],
    "data_stream.type": [
      "logs"
    ],
    "process.parent.args_count": [
      0
    ],
    "host.architecture": [
      "x86_64"
    ],
    "process.name": [
      "python3"
    ],
    "agent.id": [
      "7bc5ce5c-de4a-4b79-81fa-45198c4d75b7"
    ],
    "process.parent.executable.text": [
      "/bin/bash"
    ],
    "ecs.version": [
      "1.11.0"
    ],
    "event.created": [
      "2022-03-30T14:36:38.267Z"
    ],
    "agent.version": [
      "8.2.0-SNAPSHOT"
    ],
    "host.os.family": [
      "ubuntu"
    ],
    "process.command_line.text": [
      "python3 -c import pty; pty.spawn(\"/bin/sh\")"
    ],
    "group.id": [
      "1023"
    ],
    "process.entity_id": [
      "N2JjNWNlNWMtZGU0YS00Yjc5LTgxZmEtNDUxOThjNGQ3NWI3LTEwMjg1LTEzMjkzMTI0NTkxLjUwMDAwMDAwMA=="
    ],
    "process.parent.name.text": [
      "bash"
    ],
    "host.ip": [
      "127.0.0.1",
      "::1",
      "10.142.0.52",
      "fe80::4001:aff:fe8e:34"
    ],
    "event.sequence": [
      62669
    ],
    "process.executable.caseless": [
      "/usr/bin/python3"
    ],
    "agent.type": [
      "endpoint"
    ],
    "process.executable.text": [
      "/usr/bin/python3"
    ],
    "event.module": [
      "endpoint"
    ],
    "host.os.kernel": [
      "5.13.0-1019-gcp #23~20.04.1-Ubuntu SMP Mon Mar 7 13:39:50 UTC 2022"
    ],
    "host.os.full.caseless": [
      "ubuntu 20.04.4"
    ],
    "process.name.caseless": [
      "python3"
    ],
    "host.id": [
      "c78e287bbb0a5328e28b435c97de1cd0"
    ],
    "process.exit_code": [
      0
    ],
    "process.executable": [
      "/usr/bin/python3"
    ],
    "process.parent.name.caseless": [
      "bash"
    ],
    "process.parent.executable.caseless": [
      "/bin/bash"
    ],
    "process.parent.executable": [
      "/bin/bash"
    ],
    "process.parent.command_line.text": [
      "-bash"
    ],
    "process.args_count": [
      0
    ],
    "data_stream.namespace": [
      "default"
    ],
    "process.args": [
      "python3",
      "-c",
      "import pty; pty.spawn(\"/bin/sh\")"
    ],
    "message": [
      "Endpoint process event"
    ],
    "host.os.Ext.variant": [
      "Ubuntu"
    ],
    "process.parent.args": [
      "-bash"
    ],
    "group.Ext.real.id": [
      "1023"
    ],
    "event.action": [
      "end"
    ],
    "event.ingested": [
      "2022-03-30T14:36:49.000Z"
    ],
    "@timestamp": [
      "2022-03-30T14:36:38.267Z"
    ],
    "host.os.platform": [
      "ubuntu"
    ],
    "process.parent.command_line.caseless": [
      "-bash"
    ],
    "event.type": [
      "end"
    ],
    "data_stream.dataset": [
      "endpoint.events.process"
    ],
    "process.command_line": [
      "python3 -c import pty; pty.spawn(\"/bin/sh\")"
    ],
    "process.hash.sha1": [
      "9b6d3f9d2129510da043a88f23c1d7dafdf8d104"
    ],
    "event.id": [
      "MYFVxCublyiVcrVQ++++01D+"
    ],
    "host.os.name.caseless": [
      "linux"
    ],
    "event.dataset": [
      "endpoint.events.process"
    ]
  }
}
</details>

**Comment by terrancedejesus** [comment]
## Interactive Terminal Spawned via Python

### Comprehensive Timeline Query
Screen Shot 2022-03-30 at 11 04 02 AM

### Timeline Added to Duplicate Rule
Screen Shot 2022-03-30 at 11 14 00 AM

**Comment by terrancedejesus** [comment]
## Findings/Notes/Questions
### Build only checks for Generic
Looks like the builds fail if you try to use a comprehensive timeline template. 
<details>
marshmallow.exceptions.ValidationError: {'rule': [ValidationError({'timeline_id': ['Must be one of: db366523-f1c6-4c1f-8731-6ce5ed9e5717, 91832785-286d-4ebe-b884-1a208d111a70, 76e52245-7519-4251-91ab-262fb1a1728c, 495ad7a7-316e-4544-8a0f-9c098daee76e.'], 'timeline_title': ['Must be one of: Generic Endpoint Timeline, Generic Network Timeline, Generic Process Timeline, Generic Threat Match Timeline.']}), ValidationError({'timeline_id': ['Must be one of: db366523-f1c6-4c1f-8731-6ce5ed9e5717, 91832785-286d-4ebe-b884-1a208d111a70, 76e52245-7519-4251-91ab-262fb1a1728c, 495ad7a7-316e-4544-8a0f-9c098daee76e.'], 'threshold': ['Missing data for required field.'], 'language': ['Must be one of: kuery, lucene.'], 'timeline_title': ['Must be one of: Generic Endpoint Timeline, Generic Network Timeline, Generic Process Timeline, Generic Threat Match Timeline.'], 'type': ['Must be equal to threshold.']}), ValidationError({'timeline_id': ['Must be one of: db366523-f1c6-4c1f-8731-6ce5ed9e5717, 91832785-286d-4ebe-b884-1a208d111a70, 76e52245-7519-4251-91ab-262fb1a1728c, 495ad7a7-316e-4544-8a0f-9c098daee76e.'], 'threat_index': ['Missing data for required field.'], 'language': ['Must be one of: kuery, lucene.'], 'threat_mapping': ['Missing data for required field.'], 'timeline_title': ['Must be one of: Generic Endpoint Timeline, Generic Network Timeline, Generic Process Timeline, Generic Threat Match Timeline.'], 'type': ['Must be equal to threat_match.']}), ValidationError({'timeline_id': ['Must be one of: db366523-f1c6-4c1f-8731-6ce5ed9e5717, 91832785-286d-4ebe-b884-1a208d111a70, 76e52245-7519-4251-91ab-262fb1a1728c, 495ad7a7-316e-4544-8a0f-9c098daee76e.'], 'machine_learning_job_id': ['Missing data for required field.'], 'timeline_title': ['Must be one of: Generic Endpoint Timeline, Generic Network Timeline, Generic Process Timeline, Generic Threat Match Timeline.'], 'type': ['Must be equal to machine_learning.'], 'anomaly_threshold': ['Missing data for required field.'], 'index': ['Unknown field.'], 'query': ['Unknown field.'], 'language': ['Unknown field.']}), ValidationError({'timeline_id': ['Must be one of: db366523-f1c6-4c1f-8731-6ce5ed9e5717, 91832785-286d-4ebe-b884-1a208d111a70, 76e52245-7519-4251-91ab-262fb1a1728c, 495ad7a7-316e-4544-8a0f-9c098daee76e.'], 'language': ['Must be one of: kuery, lucene.'], 'timeline_title': ['Must be one of: Generic Endpoint Timeline, Generic Network Timeline, Generic Process Timeline, Generic Threat Match Timeline.'], 'type': ['Must be equal to query.']})]}
</details>

### Missing Generic File Timeline
Looks like we are missing a Generic File Timeline template or maybe that was intentional?

### How to check if template fits for the rule
The best way to verify this is to use Security > Timelines > Templates > Select your template and then add the additional parameters to the query to see if the document shows up.
Screen Shot 2022-03-30 at 11 04 02 AM

**Comment by terrancedejesus** [comment]
### Suspicious CertUtil Commands
## Generic Process Timeline Template Check
Screen Shot 2022-03-30 at 11 56 33 AM

**Comment by terrancedejesus** [comment]
## RDP (Remote Desktop Protocol) from the Internet
### Generic Network Timeline Template Check
Screen Shot 2022-03-30 at 12 06 23 PM

**Comment by terrancedejesus** [comment]
## Adding Hidden File Attribute via Attrib
### Generic Process Timeline Template Check
Screen Shot 2022-03-30 at 12 19 08 PM

**Comment by terrancedejesus** [comment]
### Telnet Port Activity
## Generic Network Timeline Template Check
TBD

**Comment by terrancedejesus** [comment]
## ImageLoad via Windows Update Auto Update Client
### Generic Process Timeline Template Check
Screen Shot 2022-03-30 at 3 29 45 PM

**Comment by brokensound77** [comment]
>Looks like the builds fail if you try to use a comprehensive timeline template

Just update this with the new comprehensive template IDs and names and it will pass validation 

https://github.com/elastic/detection-rules/blob/a3d7427d291bfdf01afd6b872b69f7d46a9cbf16/detection_rules/schemas/definitions.py#L40-L45

File: rules/linux/execution_python_tty_shell.toml [hunk 1]
```diff
@@ -18,11 +20,14 @@ risk_score = 73
...
 tags = ["Elastic", "Host", "Linux", "Threat Detection", "Execution"]
+timeline_id = "e70679c2-6cde-4510-9764-4823df18f7db"
+timeline_title = "Comprehensive Process Timeline"
 timestamp_override = "event.ingested"
 type = "query"
 
 query = '''
-event.category:process and event.type:(start or process_started) and process.name:python and
+event.category:process and event.type:(start or process_started) and 
+  process.name: (python or python3) and
```
Comment (brokensound77) [inline rules/linux/execution_python_tty_shell.toml:30]: might as well make it even more compatible

```suggestion
  process.name:python* and
```

File: rules/windows/execution_command_shell_started_by_svchost.toml [hunk 2]
```diff
@@ -15,6 +17,8 @@ risk_score = 21
 rule_id = "fd7a6052-58fa-4397-93c3-4795249ccfa2"
 severity = "low"
 tags = ["Elastic", "Host", "Windows", "Threat Detection", "Execution"]
+timeline_id = "e70679c2-6cde-4510-9764-4823df18f7db"
+timeline_title = "e70679c2-6cde-4510-9764-4823df18f7db"
```
Comment (terrancedejesus) [inline rules/windows/execution_command_shell_started_by_svchost.toml:21]:
```suggestion
timeline_title = "Comprehensive Process Timeline"
```

**Review by w0rk3r — APPROVED** [review]

**Pull request merged**